esutil/bulk_indexer.go

```diff
@@ -0,0 +1,509 @@
+//
+type BulkIndexerResponse struct {
+	Took      int                                  `json:"took"`
+	HasErrors bool                                 `json:"errors"`
+	Items     []map[string]BulkIndexerResponseItem `json:"items,omitempty"`
+}
+
+// BulkIndexerResponseItem represents the Elasticsearch response item.
+//
+type BulkIndexerResponseItem struct {
+	Index      string `json:"_index"`
+	DocumentID string `json:"_id"`
+	Version    int64  `json:"_version,omitempty"`
+	Result     string `json:"result,omitempty"`
+	Status     int    `json:"status,omitempty"`
+	SeqNo      int64  `json:"_seq_no,omitempty"`
+	PrimTerm   int64  `json:"_primary_term,omitempty"`
+
+	Shards struct {
+		Total      int `json:"total,omitempty"`
+		Successful int `json:"successful,omitempty"`
+		Failed     int `json:"failed,omitempty"`
+	} `json:"_shards,omitempty"`
+
+	Error struct {
+		Type   string `json:"type"`
+		Reason string `json:"reason"`
+		Cause  struct {
+			Type   string `json:"type"`
+			Reason string `json:"reason"`
+		} `json:"caused_by"`
+	} `json:"error,omitempty"`
+}
+
+// BulkResponseJSONDecoder defines the interface for custom JSON decoders.
+//
+type BulkResponseJSONDecoder interface {
+	UnmarshalFromReader(io.Reader, *BulkIndexerResponse) error
+}
+
+// BulkIndexerDebugLogger defines the interface for a debugging logger.
+//
+type BulkIndexerDebugLogger interface {
+	Printf(string, ...interface{})
+}
+
+type bulkIndexer struct {
+	wg      sync.WaitGroup
+	queue   chan BulkIndexerItem
+	workers []*worker
+	ticker  *time.Ticker
+	stats   *bulkIndexerStats
+
+	config BulkIndexerConfig
+}
+
+type bulkIndexerStats struct {
+	numAdded    uint64
+	numFlushed  uint64
+	numFailed   uint64
+	numIndexed  uint64
+	numCreated  uint64
+	numUpdated  uint64
+	numDeleted  uint64
+	numRequests uint64
+}
+
+// NewBulkIndexer creates a new bulk indexer.
+//
+func NewBulkIndexer(cfg BulkIndexerConfig) (BulkIndexer, error) {
+	if cfg.Client == nil {
+		cfg.Client, _ = elasticsearch.NewDefaultClient()
+	}
+
+	if cfg.Decoder == nil {
+		cfg.Decoder = defaultJSONDecoder{}
+	}
+
+	if cfg.NumWorkers == 0 {
+		cfg.NumWorkers = runtime.NumCPU()
+	}
+
+	if cfg.FlushBytes == 0 {
+		cfg.FlushBytes = 5e+6
+	}
+
+	if cfg.FlushInterval == 0 {
+		cfg.FlushInterval = 10 * time.Second
+	}
+
+	bi := bulkIndexer{
+		config: cfg,
+		stats:  &bulkIndexerStats{},
+	}
+
+	bi.init()
+
+	return &bi, nil
+}
+
+// Add adds an item to the indexer.
+//
+func (bi *bulkIndexer) Add(ctx context.Context, item BulkIndexerItem) error {
+	atomic.AddUint64(&bi.stats.numAdded, 1)
+
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case bi.queue <- item:
+	}
+
+	return nil
+}
+
+// Close calls flush on writers and closes the indexer queue channel.
+//
+func (bi *bulkIndexer) Close(ctx context.Context) error {
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	default:
+		close(bi.queue)
+		bi.wg.Wait()
+	}
+
+	bi.ticker.Stop()
+
+	for _, w := range bi.workers {
```

Comment:
Rather than flushing here, how about having the workers flush upon seeing their channel is closed, but before signalling that they're done?